src/uct/base/uct_md.c

```diff
@@ -250,43 +250,47 @@ static ucs_status_t uct_config_read(uct_config_bundle_t **bundle,
 }
 
 static uct_tl_component_t *uct_find_tl_on_md(uct_md_component_t *mdc,
+                                             uint64_t md_flags,
                                              const char *tl_name)
 {
     uct_md_registered_tl_t *tlr;
 
-    ucs_list_for_each(tlr, &mdc->tl_list, list) {
-        if (!strcmp(tl_name, tlr->tl->name)) {
-            return tlr->tl;
+    if ((tl_name == NULL) && (md_flags & UCT_MD_FLAG_SOCKADDR)) {
+        tlr = ucs_list_head(&mdc->tl_list, uct_md_registered_tl_t, list);
+        if (tlr == NULL) {
+            return NULL;
         }
-    }
-    return NULL;
-}
-
-static uct_tl_component_t *uct_find_tl(const char *tl_name)
-{
-    uct_md_component_t *mdc;
-    uct_tl_component_t *tlc;
-
-    ucs_list_for_each(mdc, &uct_md_components_list, list) {
-        tlc = uct_find_tl_on_md(mdc, tl_name);
-        if (tlc != NULL) {
-            return tlc;
+        return tlr->tl;
+    } else {
+        if (tl_name != NULL) {
+            ucs_list_for_each(tlr, &mdc->tl_list, list) {
+                if (!strcmp(tl_name, tlr->tl->name)) {
+                    return tlr->tl;
+                }
+            }
         }
+        return NULL;
     }
-    return NULL;
 }
 
-ucs_status_t uct_iface_config_read(const char *tl_name, const char *env_prefix,
-                                   const char *filename,
-                                   uct_iface_config_t **config_p)
+ucs_status_t uct_md_iface_config_read(uct_md_h md, const char *tl_name,
+                                      const char *env_prefix, const char *filename,
+                                      uct_iface_config_t **config_p)
 {
     uct_config_bundle_t *bundle = NULL;
     uct_tl_component_t *tlc;
+    uct_md_attr_t md_attr;
     ucs_status_t status;
 
-    tlc = uct_find_tl(tl_name);
+    md->ops->query(md, &md_attr);
+    tlc = uct_find_tl_on_md(md->component, md_attr.cap.flags, tl_name);
+
     if (tlc == NULL) {
-        ucs_error("Transport '%s' does not exist", tl_name);
+        if ((tl_name == NULL) && (md_attr.cap.flags & UCT_MD_FLAG_SOCKADDR)) {
```

Comment:
no need to check for UCT_MD_FLAG_SOCKADDR here. if user provides tl_name=NULL and the md does not support sockaddr, it would go to second case and print "Transport (null) does not exist" which is not the right behavior.

Reply:
removing it will result in - "There is no sockaddr transport registered on the md" which isn't correct as well. 
"Transport (null) does not exist" was what was printed before this change for NULL in tl_name (while NULL isn't allowed now as well for DEVICE open mode).

Reply:
but "There is no sockaddr transport registered on the md" is the correct description, in case md does not support sockaddr

Reply:
if the user passed a null as the tl_name and the md doesnt support UCT_MD_FLAG_SOCKADDR, why print a sockaddr related message?

Reply:
because when the user passes a null as tl_name, it means the user intention is to read a sockaddr configuration.
if this fails, need to print a sockaddr related message.